syntax = "proto3";
package stats;

enum MetricKind {
  VIEWS = 0;
  LIKES = 1;
  COMMENTS = 2;
}

message PostIdRequest {
  string post_id = 1;
}

message PostCountersResponse {
  int64 views = 1;
  int64 likes = 2;
  int64 comments = 3;
}

message TimelinePoint {
  string date = 1;
  int64  value = 2;
}

message TimelineRequest {
  string post_id = 1;
  MetricKind metric = 2;
}

message TimelineResponse {
  repeated TimelinePoint points = 1;
}

message TopPostsRequest {
  MetricKind metric = 1;
}

message TopPost {
  string post_id = 1;
  int64  value = 2;
}

message TopPostsResponse {
  repeated TopPost posts = 1;
}

message TopUsersRequest {
  MetricKind metric = 1;
}

message TopUser {
  string user_id = 1;
  int64  value = 2;
}

message TopUsersResponse {
  repeated TopUser users = 1;
}

service StatsService {
  rpc GetPostCounters(PostIdRequest)          returns (PostCountersResponse);
  rpc GetPostTimeline(TimelineRequest)        returns (TimelineResponse);
  rpc GetTopPosts(TopPostsRequest)            returns (TopPostsResponse);
  rpc GetTopUsers(TopUsersRequest)            returns (TopUsersResponse);
}
